kyaml/setters2/set.go

```diff
@@ -155,27 +163,54 @@ func (s *Set) substitute(field *yaml.RNode, ext *cliExtension, _ *spec.Schema) (
 }
 
 // set applies the value from ext to field if its name matches s.Name
-func (s *Set) set(field *yaml.RNode, ext *cliExtension, sch *spec.Schema) bool {
+func (s *Set) set(field *yaml.RNode, ext *cliExtension, sch *spec.Schema) (bool, error) {
 	// check full setter
 	if ext.Setter == nil || !s.isMatch(ext.Setter.Name) {
-		return false
+		return false, nil
 	}
 
-	// TODO(pwittrock): validate the field value
+	if err := validateAgainstSchema(ext, sch); err != nil {
+		return false, err
+	}
 
 	if val, found := ext.Setter.EnumValues[ext.Setter.Value]; found {
 		// the setter has an enum-map.  we should replace the marker with the
 		// enum value looked up from the map rather than the enum key
 		field.YNode().Value = val
-		return true
+		return true, nil
 	}
 
 	// this has a full setter, set its value
 	field.YNode().Value = ext.Setter.Value
 
 	// format the node so it is quoted if it is a string
 	yaml.FormatNonStringStyle(field.YNode(), *sch)
-	return true
+	return true, nil
+}
+
+// validateAgainstSchema validates the input setter value against user provided
+// openAI schema
+func validateAgainstSchema(ext *cliExtension, sch *spec.Schema) error {
+	sc := spec.Schema{}
+	sc.Properties = make(map[string]spec.Schema)
+	sc.Properties[ext.Setter.Name] = *sch
+
+	input := map[string]interface{}{}
+	format := `{"%s" : "%s"}`
+	if yaml.IsValueNonString(ext.Setter.Value) {
+		format = `{"%s" : %s}`
+	}
+
+	inputJSON := fmt.Sprintf(format, ext.Setter.Name, ext.Setter.Value)
+	err := json.Unmarshal([]byte(inputJSON), &input)
```

Comment:
I might be wrong here, but it seems like this is a lot of code for just creating a map with one key-value pair in it? Do we need to create the json string and then parse it?

Reply:
To take advantage of json.Unmarshal. So the problem is, if publisher set required type to be integer and user inputs 4 as value, cli tool takes it as string. We need a way to say that 4 is an integer and not string, similarly true is boolean and not string("true"). And the unmarshalled schema is fed into validate.AgainstSchema. I followed this from their test case. https://github.com/go-openapi/validate/blob/master/doc_test.go#L145